newton/utils/render.py

```diff
@@ -464,9 +506,193 @@ class SimRendererUsd(CreateSimRenderer(renderer=UsdRenderer)):
+            for prim_path, body_id in self.path_body_map.items():
+                full_xform = body_q[body_id]
+                # TODO: do this once in __init__
+                # TODO: sanity check this with Eric Heiden
+                # Take relative xform into account
+                rel_xform = self.path_body_relative_transform.get(prim_path)
+                if rel_xform:
+                    full_xform = wp.mul(full_xform, rel_xform)
+
+                full_xform = self._apply_parents_inverse_xform(full_xform, prim_path)
+                self._update_usd_prim_xform(prim_path, full_xform, sim_time)
+
+
+    def render(
+        self,
+        state: newton.State,
+        sim_time,
+    ):
+        self.render_time_sampled_transform(state, sim_time)
+
+    def _apply_parents_inverse_xform(self, full_xform, prim_path):
+        """
+        Transformation in Warp sim consists of translation and pure rotation: trnslt and quat.
+        Transformations of bodies are stored in body_q in simulation state.
+        For sim_usd, trnslt is computed directly from PhysicsUtils by the function GetRigidBodyTransformation
+        in parseUtils.cpp:
+            const GfMatrix4d mat = UsdGeomXformable(bodyPrim).ComputeLocalToWorldTransform(UsdTimeCode::Default());
+            const GfTransform tr(mat);
+            const GfVec3d pos = tr.GetTranslation();
+            const GfQuatd rot = tr.GetRotation().GetQuat();
+        In import_nvusd, we set trnslt = pos and quat = fromgfquat(rot), where fromgfquat has the following logic:
+        wp.normalize(wp.quat(*gfquat.imaginary, gfquat.real)).
+
+        For trnslt, we have:
+            warp_trnslt = xform.ComputeLocalToWorldTransform().GetTranslation().
+        But in USD space:
+            xform.ComputeLocalToworldTransform() = xform.GetLocalTransform() * xform.ComputeParentToWorldTransform()
+            Prim_LTW_USD = Prim_Local_USD * Parent_LTW_USD,
+        or in Warp space, we work with transpose and arrive at:
+            Prim_LTW = Parent_LTW * Prim_Local
+            warp_trnslt = p_Rot * prim_trnslt + p_trnslt,
+            i.e.
+            prim_trnslt = p_inv_Rot * (warp_trnslt - p_trnslt).
+
+        For rotation, we have:
+            rot = tr.GetRotation().GetQuat();
+            warp_quat = wp.normalize(wp.quat(rot)).
+        However, rot is already normalized, so we don't actually need to renormalize it.
+        So in Warp space,
+            warp_Rot = p_Rot * diag(1/s_x, 1/s_y, 1/s_z) * prim_Rot, so
+            prim_Rot = wp.inv(p_Rot * diag(1/s_x, 1/s_y, 1/s_z)) * warp_Rot
+
+        Both p_inv_Rot and wp.inv(p_Rot * diag(1/s_x, 1/s_y, 1/s_z)) do not change during sim, so they are computed in __init__.
+        """
+        c_prim = self.stage.GetPrimAtPath(Sdf.Path(prim_path))
+        p_path = str(c_prim.GetParent().GetPath())
+
+        if p_path in self.builder_results["path_body_map"]:
+            return
+
+        p_trnslt = self.p_trnslts[p_path]
```

Comment:
I'm not sure if you initialized these members in the constructor? Also the naming should be clearer to match the other member variables: e.g. `path_translations` or `path_positions`.

Reply:
Fixed in the latest iteration of this PR.